src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -1852,6 +1851,96 @@ let BuildFieldMap (cenv: cenv) env isPartial ty (flds: ((Ident list * Ident) * '
+                let reducedRecd = (lidwid, Some(SynExpr.Record (baseInfo, copyInfo, aFlds @ bFlds, m))) 
+                groupIfNested (reducedRecd :: res) ys
+            | (lidwid, Some (SynExpr.AnonRecd (isStruct, copyInfo, aFlds, m, trivia))), (_, Some (SynExpr.AnonRecd (recordFields = bFlds))) ->
+                let reducedRecd = (lidwid, Some(SynExpr.AnonRecd (isStruct, copyInfo, aFlds @ bFlds, m, trivia))) 
+                groupIfNested (reducedRecd :: res) ys
+            | _ -> groupIfNested (x :: res) (y :: ys)
+
+    fields
+    |> List.groupBy (fun ((_, field), _) -> field.idText)
+    |> List.collect (fun (_, fields) ->
+        if fields.Length < 2 then 
+            fields
+        else
+            groupIfNested [] fields)
+
+/// Expands a long identifier into nested copy-and-update expressions
+let TransformAstForNestedUpdates cenv env overallTy (lid: LongIdent) exprBeingAssigned withExpr =
+    let recdExprCopyInfo ids withExpr id =
+        let upToId origSepRng id lidwd =
+            let rec buildLid res (id: Ident) =
+                function
+                | [] -> res
+                | (h: Ident) :: t ->
+                    // Mark these hidden field accesses as synthetic so that they don't make it
+                    // into the name resolution sink.
+                    let h = ident(h.idText, h.idRange.MakeSynthetic())
+                    if equals h.idRange id.idRange then h :: res else buildLid (h :: res) id t
+
+            let calcLidSeparatorRanges origSepRng lid =
+                match lid with
+                | [] | [_] -> [ origSepRng ]
+                | _ :: t -> origSepRng :: List.map (fun (s: Ident, e: Ident) -> mkRange s.idRange.FileName s.idRange.End e.idRange.Start) t
+
+            let lid = buildLid [] id lidwd |> List.rev
+
+            (lid, List.pairwise lid |> calcLidSeparatorRanges origSepRng)
+
+        let totalRange (origId: Ident) (id: Ident) =
+            mkRange origId.idRange.FileName origId.idRange.End id.idRange.Start
+
+        let rangeOfBlockSeperator (id: Ident) =
+            let idEnd = id.idRange.End
+            let blockSeperatorStartCol = idEnd.Column
+            let blockSeperatorEndCol = blockSeperatorStartCol + 4
+            let blockSeperatorStartPos = mkPos idEnd.Line blockSeperatorStartCol
+            let blockSeporatorEndPos = mkPos idEnd.Line blockSeperatorEndCol
+
+            mkRange id.idRange.FileName blockSeperatorStartPos blockSeporatorEndPos
+
+        match withExpr with 
+        | SynExpr.Ident origId, (sepRange, _) ->
+            let lid, rng = upToId sepRange id (origId :: ids)
+            Some (SynExpr.LongIdent (false, LongIdentWithDots(lid, rng), None, totalRange origId id), (rangeOfBlockSeperator id, None))
+        | _ -> None
+
+    let rec synExprRecd copyInfo (id: Ident) fields exprBeingAssigned =
+        match fields with 
+        | [] -> failwith "unreachable"
```

Comment:
Famous last words :D

Reply:
I use this in a couple of places. The question is whether we need to fail more gracefully. I'd rather do this than return some form of empty result that would lead to undefined behavior in the future, should some hypothetical parser changes trigger enable these code paths.